nifi-nar-bundles/nifi-email-bundle/nifi-email-processors/src/main/java/org/apache/nifi/processors/email/ConsumeEWS.java

```diff
@@ -0,0 +1,518 @@
+
+                SearchFilter sf = new SearchFilter.SearchFilterCollection(LogicalOperator.And, new SearchFilter.IsEqualTo(EmailMessageSchema.IsRead, false));
+                FindItemsResults<Item> findResults = service.findItems(folder.getId(), sf, view);
+
+                if(findResults == null || findResults.getItems().size()== 0){
+                    return;
+                }
+
+                service.loadPropertiesForItems(findResults, PropertySet.FirstClassProperties);
+
+                for (Item item : findResults) {
+                    EmailMessage ewsMessage = (EmailMessage) item;
+                    messageQueue.add(parseMessage(ewsMessage));
+
+                    if(deleteOnRead){
+                        ewsMessage.delete(DeleteMode.HardDelete);
+                    } else if(markAsRead){
+                        ewsMessage.setIsRead(true);
+                        ewsMessage.update(ConflictResolutionMode.AlwaysOverwrite);
+                    }
+                }
+
+                service.close();
+            } catch (Exception e) {
+                throw new ProcessException("Failed retrieving new messages from EWS.", e);
+            }
+        }
+    }
+
+    protected Folder getFolder(ExchangeService service) {
+        Folder folder;
+        if(folderName.equals("INBOX")){
+            try {
+                folder = Folder.bind(service, WellKnownFolderName.Inbox);
+            } catch (Exception e) {
+                throw new ProcessException("Failed to bind Inbox Folder on EWS Server", e);
+            }
+        } else {
+            FolderView view = new FolderView(10);
+            view.setTraversal(FolderTraversal.Deep);
+            SearchFilter searchFilter = new SearchFilter.IsEqualTo(FolderSchema.DisplayName, folderName);
+            try {
+                FindFoldersResults foldersResults = service.findFolders(WellKnownFolderName.Root,searchFilter, view);
+                ArrayList<Folder> folderIds = foldersResults.getFolders();
+                if(folderIds.size() > 1){
+                    throw new ProcessException("More than 1 folder found with the name " + folderName);
+                }
+
+                folder = Folder.bind(service, folderIds.get(0).getId());
+            } catch (Exception e) {
+                throw new ProcessException("Search for Inbox Subfolder failed.", e);
+            }
+        }
+        return folder;
+    }
+
+    public MimeMessage parseMessage(EmailMessage item) throws Exception {
+        EmailMessage ewsMessage = item;
+        final String bodyText = ewsMessage.getBody().toString();
+
+        MultiPartEmail mm;
+
+        if(ewsMessage.getBody().getBodyType() == BodyType.HTML){
+            mm = new HtmlEmail().setHtmlMsg(bodyText);
+        } else {
+            mm = new MultiPartEmail();
+            mm.setMsg(bodyText);
+        }
+        mm.setHostName("NiFi-EWS");
```

Comment:
What's really going on here is I'm converting message formats, and the API for Apache Email requires you to have a HostName as the originator of the message, because it thinks I'm trying to send a new email rather than creating an in memory message that will never be sent..  Unfortunately I can't just leave this field empty, otherwise the underlying library fails.